Added Dockerfile

srkiNZ84 opened the pull request:
Added a Dockerfile to build an image to run puppeteer. Have added a
configuration variable as Chromium throws an error when running as
"root" and not having the "--no-sandbox" flag there.

Updated the README with Docker instructions.

TODO: Instead of installing Chromium through the Debian repos in order
to get the dependencies, we should just install the dependencies by
themselves

TODO: Add config to create a non-root user to run as

TODO: Fix up the file permissions issue where the generated archives
have owner/group set to "root"

Issue(s): None

Commit:
Added Dockerfile

Added a Dockerfile to build an image to run puppeteer. Have added a
configuration variable as Chromium throws an error when running as
"root" and not having the "--no-sandbox" flag there.

Updated the README with Docker instructions.

TODO: Instead of installing Chromium through the Debian repos in order
to get the dependencies, we should just install the dependencies by
themselves

TODO: Add config to create a non-root user to run as

TODO: Fix up the file permissions issue where the generated archives
have owner/group set to "root"

Issue(s): None

diegocr commented on index.js:
```diff
@@ -247,7 +248,14 @@ async function getSiteVersion(page) {
     'use strict';
 
-    const browser = await puppeteer.launch();
+    var chrome_args = [];
```
```suggestion
    const args = [];
```

diegocr commented on index.js:
```diff
@@ -247,7 +248,14 @@ async function getSiteVersion(page) {
-    const browser = await puppeteer.launch();
+    var chrome_args = [];
+    if (config.DISABLE_CHROME_SANDBOX){
+      chrome_args = ['--no-sandbox'];
```
```suggestion
      args.push('--no-sandbox');
```

srkiNZ84 replied:
Thanks for this, I was not aware of the "push" function to add elements to a const :+1:

diegocr commented on index.js:
```diff
@@ -247,7 +248,14 @@ async function getSiteVersion(page) {
+      chrome_args = ['--no-sandbox'];
+    }
+
+    const browser = await puppeteer.launch({
+       args: chrome_args,
+    }).catch(console.error);
```
If this does fail we shouldn't continue by printing the error to the console, i think we rather halt the execution.

i.e.

```suggestion
    const browser = await puppeteer.launch({args});
```

srkiNZ84 replied:
I think this was an attempt to get some information around why Chromium wasn't launching successfully as the execution fails anyway whether we catch the error or not.

diegocr commented on Dockerfile:
```diff
@@ -0,0 +1,15 @@
+RUN apt-get install -y nodejs
+
+# NOTE: We are only installing chromium to get its dependencies
+RUN apt install -y chromium
```
What dependencies exactly, and won't those cause any issue by using an older chromium version?

i.e. given puppeteer v5.3.1 comes with Chromium 86

srkiNZ84 replied:
> What dependecies exactly

On Linux the Chromium package depends on a whole bunch of OS libraries. Specifically:
```
# apt-cache depends chromium
chromium
  Depends: libasound2
  Depends: libatk-bridge2.0-0
  Depends: libatk1.0-0
  Depends: libatomic1
  Depends: libatspi2.0-0
  Depends: libc6
  Depends: libcairo2
  Depends: libcups2
  Depends: libdbus-1-3
  Depends: libdrm2
  Depends: libevent-2.1-7
  Depends: libexpat1
  Depends: libflac8
  Depends: libfontconfig1
  Depends: libfreetype6
  Depends: libgbm1
  Depends: libgcc-s1
  Depends: libglib2.0-0
  Depends: libjpeg62-turbo
  Depends: libjsoncpp24
  Depends: liblcms2-2
  Depends: libminizip1
  Depends: libnspr4
  Depends: libnss3
  Depends: libopenjp2-7
  Depends: libopus0
  Depends: libpango-1.0-0
  Depends: libpng16-16
  Depends: libpulse0
  Depends: libre2-9
  Depends: libsnappy1v5
  Depends: libstdc++6
  Depends: libwebp6
  Depends: libwebpdemux2
  Depends: libwebpmux3
  Depends: libx11-6
  Depends: libxcb1
  Depends: libxcomposite1
  Depends: libxdamage1
  Depends: libxext6
  Depends: libxfixes3
  Depends: libxkbcommon0
  Depends: libxml2
  Depends: libxrandr2
  Depends: libxslt1.1
  Depends: zlib1g
```
The Chromium shipped with puppeteer doesn't install these (assumes that these are there) and fails to launch if they are missing in the docker container. 

> won't those cause any issue by using an older chromium version?

From the list above as you can see these provide fairly standard functionality, mostly around parsing image formats and XML etc... they are fairly stable and while there might have been changes from version 86 to 103, I have tested running the puppeteer code and can confirm that Chromium 86 launches successfully with these dependencies installed.

I'll replace the above "install chromium" with "install this list of packages".

diegocr replied:
Thanks, let's see how this goes in the long run :)

diegocr commented on README.md:
```diff
@@ -74,6 +74,19 @@ This will run the app in archive-only mode and exit, creating gzipped bundles as
+    docker run -e DISABLE_CHROME_SANDBOX=true -v /home/usre/code/blofetcha/docker-archive:/archive blofetcha
+
+NOTE: The mounted "docker-archive" volume will contain the archived Javascript files
+NOTE2: Currently the files are output as having `root:root` permissions and need to be changed after
```
I would remove this line if it isn't something we can do about, and if it is we should just do it (?) so either way let's get rid of it, and then remove the `NOTE:` prefix in the previous one <\/nit> :)

diegocr reviewed (changes requested):
Thanks Serge! a few comments below :)

Commit:
Change chrome args to const

For the sake of consistency, change the chromium args to be a const
rather than a var and use the "push()" function to add the values to the
array when needed.

Issue(s): None

Commit:
Explicitly list Chromium dependencies

Rather than installing the system provided Chromium and getting the
dependencies that way, list them out and install them so that the
Chromium browser bundled with puppeteer can launch.

Issue(s): None

Commit:
Removed note about permissions from README

There is no nice, straight forward solution wrt Docker and file
permissions between the host and the docker container. As such we
can assume that the person running the container is capable of figuring
out and changing the permissions of the archive files.

Issue(s): None

diegocr reviewed (approved):
LGTM now :)

Thanks!
